ktor-server/jvm/src/io/ktor/server/application/plugins/api/ServerPlugin.kt

```diff
@@ -12,31 +12,56 @@ import io.ktor.server.request.*
 import io.ktor.server.response.*
 import io.ktor.util.*
 import io.ktor.util.pipeline.*
-import kotlin.random.Random
+import kotlin.random.*
 
 /**
  * A plugin that embeds into the HTTP pipeline and extends Ktor functionality.
  **/
-public abstract class ServerPlugin<Configuration : Any> private constructor(
-    pluginFactory: ServerPluginFactory<Configuration>
+public sealed class ServerPlugin<Configuration : Any> private constructor(
+    internal val key: AttributeKey<out ServerPlugin<Configuration>>
```

Comment:
Please, add `key` to KDoc.
Also, I'm not sure it's a good idea to require `key` from a user. `AttributeKey` seems to me as some weared wrapper class, and we decided to move to simpler abstractions.
Maybe receive `name: String` here and then construct a `key` from it? Same way it is now done in `createPlugin`

Reply:
This constructor is private and invoked only inside `creade[X]Plugin`. So users can not pass `key` here.